core/primitives/src/congestion_info.rs

```diff
@@ -0,0 +1,103 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use bytesize::ByteSize;
+use near_primitives_core::types::Gas;
+
+/// Stores the congestion level of a shard.
+///
+/// [`CongestionInfo`] should remain an internal struct that is not borsh
+/// serialized anywhere. This way we can change it more easily.
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
+pub struct CongestionInfo {
+    /// Congestion level for queued up unprocessed receipts.
+    pub incoming_congestion: u16,
+    /// Congestion level for queued up unprocessed receipts and buffered
+    /// outgoing receipts.
+    pub memory_congestion: u16,
+}
+
+/// Detailed information about receipts currently stored in the trie,
+/// potentially contributing to congestion.
+///
+/// This data is stored in the [`ChunkExtra`], therefore changing its layout
+/// requires a protocol change.
+#[derive(Default, Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+pub struct StoredReceiptsInfo {
+    /// Action execution gas plus attached gas of all receipts in the delayed
+    /// receipts queue.
+    pub delayed_receipts_gas: Gas,
+    /// Size of borsh serialized receipts in the delayed receipts queue.
+    pub delayed_receipts_bytes: u64,
+    /// Size of borsh serialized postponed receipts.
+    pub postponed_receipts_bytes: u64,
+    /// Size of borsh serialized receipts in the outgoing buffer.
+    pub buffered_receipts_bytes: u64,
+}
+
+impl StoredReceiptsInfo {
+    pub fn nep539_congestion_info(&self) -> CongestionInfo {
```

Comment:
nit: Can we avoid referencing the nep number in the method names? 

If you're foreseeing changes to the stored info perhaps we append struct names with V1 and put them inside an enum like we do for Block at al?